packages/inquirer/src/ui/prompt.mts

```diff
@@ -303,48 +304,60 @@ export default class PromptsRunner<A extends Answers> {
       throw new Error(`Prompt for type ${question.type} not found`);
     }
 
-    return isPromptConstructor(prompt)
-      ? defer(() => {
-          const rl = readline.createInterface(
-            setupReadlineOptions(this.opt),
-          ) as InquirerReadline;
-          rl.resume();
-
-          const onClose = () => {
-            rl.removeListener('SIGINT', this.onForceClose);
-            rl.setPrompt('');
-            rl.output.unmute();
-            rl.output.write(ansiEscapes.cursorShow);
-            rl.output.end();
-            rl.close();
-          };
-          this.onClose = onClose;
-          this.rl = rl;
-
-          // Make sure new prompt start on a newline when closing
-          process.on('exit', this.onForceClose);
-          rl.on('SIGINT', this.onForceClose);
-
-          const activePrompt = new prompt(question, rl, this.answers);
-
-          return from(
-            activePrompt.run().then((answer: unknown) => {
+    const promptFn: PromptFn = isPromptConstructor(prompt)
+      ? (q, opt = {}) =>
+          new Promise<any>((resolve, reject) => {
+            const rl = readline.createInterface(
+              setupReadlineOptions(this.opt),
+            ) as InquirerReadline;
+            rl.resume();
+
+            const onClose = () => {
+              rl.removeListener('SIGINT', this.onForceClose);
+              rl.setPrompt('');
+              rl.output.unmute();
+              rl.output.write(ansiEscapes.cursorShow);
+              rl.output.end();
+              rl.close();
+            };
+            this.onClose = onClose;
+            this.rl = rl;
+
+            // Make sure new prompt start on a newline when closing
+            process.on('exit', this.onForceClose);
+            rl.on('SIGINT', this.onForceClose);
+
+            const activePrompt = new prompt(q, rl, this.answers);
+
+            const cleanup = () => {
               onClose();
               this.onClose = undefined;
               this.rl = undefined;
+            };
+            const { signal } = opt;
+            if (signal) {
+              const abort = () => {
+                reject(new AbortPromptError({ cause: signal.reason }));
+                cleanup();
+              };
+              if (signal.aborted) {
+                abort();
+                return;
+              }
+              signal.addEventListener('abort', () => abort());
```

Comment:
I think we'll need a `removeEventListener` here too

Reply:
Done